bluetooth/src/commonMain/kotlin/UUID.kt

```diff
@@ -15,12 +15,58 @@
 
  */
 
+@file:JvmName("CommonUUID")
 package com.splendo.kaluga.bluetooth
 
+import com.splendo.kaluga.base.text.format
+import kotlin.jvm.JvmName
+
+private object Constants {
+    val longFormatValidationRegex = Regex("[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}", RegexOption.IGNORE_CASE)
+    val shortFormatValidationRegex = Regex("[0-9a-f]{4}", RegexOption.IGNORE_CASE)
+    const val baseBluetoothUUID = "0000%s-0000-1000-8000-00805f9b34fb"
+}
+
+sealed class UUIDException: Exception() {
+    class InvalidFormat(uuidString:String) : Exception("String '$uuidString' does not represent a valid UUID")
+}
+
 expect class UUID
 
 expect val UUID.uuidString: String
 
-expect fun uuidFrom(uuidString:String):UUID
+@Throws(UUIDException::class)
+fun uuidFrom(uuidString : String): UUID =
+    if(uuidString.isValidUUIDString()) unsafeUUIDFrom(uuidString)
+    else throw UUIDException.InvalidFormat(uuidString)
 
 expect fun randomUUID():UUID
+
+/**
+ * The UUID format is the same for all services and characteristics and defined in bluetooth specification.
+ *
+ * @see <a href="https://www.bluetooth.com/specifications/gatt/characteristics/">GATT Characteristics specification</a>
+ * @see <a href="https://btprodspecificationrefs.blob.core.windows.net/assigned-numbers/Assigned%20Number%20Types/Service%20Discovery.pdf">Service discovery</a>
+ */
+internal fun uuidFromShort(uuidString: String): UUID = uuidFrom(Constants.baseBluetoothUUID.format(uuidString))
+
+/**
+ * Meant for internal usage. It takes string which already passed validation
+ */
+internal expect fun unsafeUUIDFrom(uuidString: String): UUID
+
+internal fun String.isShortUUID() = length == 4
+
+internal fun String.isValidUUIDString(): Boolean =
+    Constants.longFormatValidationRegex.matches(this) ||
```

Comment:
I think it's too much to scan string twice with regex to try to create UUID, why not just make it optional, and let client to decide when to unwrap it?

Reply:
I hardly can imagine a situation when it can be a performance issue. Normally we are not creating a big amount of UUID in a loop. Of course, we can make the result of `uuidFrom(string)` optional, but in this case, we still need to make this validation to achieve similar behaviour on different platforms. Android throws an exception in format is wrong, on JS (as you may notice before) we have custom UUID implementation so it is not showing any error (if we are not validation string). 
In general, we can scan only long format and short format to validate differently, but not sure it is faster and regex gives us an easier way to texted validation (for instance if we want to consider short format with '0x' prefix as a valid). Also we can wrap android implementation in try-catch and return null in case of exception.

I actually don't mind to reimplement but first would be interesting to hear what guys think about it

Reply:
Not sure if it is more efficient, but I updated this part, now it uses only one regex to validate format